### PR TITLE
Fix Snyk code security findings

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary
- harden archive extraction and URL-derived paths in shared download helpers
- restrict dataset YAML scripts to trusted bundled dataset configs and replace unsafe eval usage
- clean up example/test findings for postMessage target origins, metadata loading, C++ ownership, label bounds, and CUDA device parsing

## Tests
- ruff check ultralytics/utils/downloads.py ultralytics/utils/checks.py ultralytics/data/utils.py ultralytics/nn/tasks.py tests/test_python.py tests/test_cuda.py examples/YOLOv8-TFLite-Python/main.py examples/YOLOv8-ONNXRuntime/main.py examples/YOLOv8-OpenCV-ONNX-Python/main.py examples/YOLOv8-Segmentation-ONNXRuntime-Python/main.py
- pytest tests/test_python.py -k "safe_download"
- pytest tests/test_cuda.py -k test_train (skipped: no CUDA devices available)
- python -m compileall -q ultralytics/utils/downloads.py ultralytics/utils/checks.py ultralytics/data/utils.py ultralytics/nn/tasks.py tests/test_python.py tests/test_cuda.py examples/YOLOv8-TFLite-Python/main.py examples/YOLOv8-ONNXRuntime/main.py examples/YOLOv8-OpenCV-ONNX-Python/main.py examples/YOLOv8-Segmentation-ONNXRuntime-Python/main.py
- git diff --check

## Notes
- snyk CLI is not installed locally, so this PR should be retested in Snyk after CI.
- bun run knip was requested but this repo has no knip script configured: it exits with "Script not found 'knip'".

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🔒 This PR strengthens file-handling security, improves example reliability, and fixes a few usability issues across docs, examples, downloads, and tests.

### 📊 Key Changes
- 🔐 **Hardened archive extraction** in `safe_download()` and `unzip_file()`:
  - Blocks unsafe ZIP and TAR entries that could extract outside the intended folder.
  - Skips symbolic links and other risky TAR members.
  - Replaces shell-based TAR extraction with safer Python-based handling.
- 🛡️ **Validated Ultralytics Platform URIs** in `check_file()`:
  - Rejects unsafe `ul://` paths containing absolute paths or parent-directory traversal.
- 🌐 **Improved iframe theme syncing** in docs:
  - `postMessage()` now sends messages to the iframe’s actual origin instead of `"*"`, making embedded theme updates safer.
- 📦 **Made download filename fallback safer**:
  - `url2file()` now returns `"download"` if a URL does not provide a valid filename.
- 🧪 **Added tests for unsafe archive handling**:
  - New test cases confirm ZIP and TAR files with malicious paths are skipped during extraction.
- 🖥️ **Cleaned up ONNX/C++ example code**:
  - Switched from heap allocation to stack allocation for `YOLO_V8` in example code.
  - Simplified function signatures from pointer references to plain pointers.
  - Improved label box sizing using actual rendered text width, avoiding oversized or clipped labels.
- 📚 **Updated Python example dataset loading**:
  - Example scripts now load dataset YAMLs directly from `ROOT / "cfg/datasets/..."` instead of using `check_yaml()`.
- ✅ **Improved TFLite metadata validation**:
  - Confirms metadata is a real YAML file before loading, with clearer error handling.
- 🎮 **Fixed CUDA test parsing**:
  - `CUDA_VISIBLE_DEVICES` is now parsed more robustly for single- and multi-GPU setups.

### 🎯 Purpose & Impact
- 🔒 **Better security for users** by preventing path traversal and unsafe archive extraction when downloading or unpacking files.
- ☁️ **Safer integration with the Ultralytics Platform** by rejecting suspicious `ul://` paths before they can create unsafe local paths.
- 🌐 **More secure docs behavior** by limiting iframe theme messages to trusted origins.
- 🧰 **More reliable example projects** with cleaner memory usage, better file loading, and improved metadata validation.
- 🎨 **Better visual output in C++ demos** thanks to label backgrounds that match real text size more accurately.
- 🧪 **Stronger test coverage** helps catch regressions in download safety and device handling.
- 👥 **Low risk for most users**, with the biggest impact being improved safety and more dependable developer examples rather than API-breaking changes.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist you shortly. Please review the checklist below to help ensure a smooth merge 🔍✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: tests/test_python.py [hunk 1]
```diff
@@ -845,7 +874,7 @@ def test_grayscale(task: str, model: str, data: str, tmp_path) -> None:
     YAML.save(data=data, file=grayscale_data)
     # remove npy files in train/val splits if exists, might be created by previous tests
     for split in {"train", "val"}:
-        for npy_file in (Path(data["path"]) / data[split]).glob("*.npy"):
+        for npy_file in Path(data[split]).glob("*.npy"):
```
Comment (UltralyticsAssistant) [inline tests/test_python.py]: 💡 **MEDIUM**: This drops the dataset root from the cleanup path. In these dataset YAMLs, `train`/`val` are typically relative paths like `images/train`, so `Path(data[split])` will look under the current working directory and leave stale `.npy` files behind. That can make this test flaky depending on prior runs.

**Suggested change:**
```suggestion
        for npy_file in (Path(data["path"]) / data[split]).glob("*.npy"):
```

File: ultralytics/nn/tasks.py [hunk 2]
```diff
@@ -1564,7 +1564,10 @@ def parse_model(d, ch, verbose=True):
         depth, width, max_channels = scales[scale]
 
     if act:
-        Conv.default_act = eval(act)  # redefine default activation, i.e. Conv.default_act = torch.nn.SiLU()
+        match = re.fullmatch(r"(?:torch\.)?nn\.([A-Za-z_]\w*)\(\)", act)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py]: 💡 **MEDIUM**: This is a backwards-compatibility regression for custom model YAMLs that use parameterized activations such as `nn.LeakyReLU(0.1)` or `nn.Hardswish(inplace=True)`. Those were previously valid because `act` was evaluated directly, but the new regex only accepts zero-argument constructors. Since `act` is part of the public config surface, this will reject existing user models unrelated to the security issue.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The security hardening changes look good overall, especially around archive extraction and path validation. I found two regressions worth fixing before merge: the grayscale test cleanup now ignores the dataset root, and the activation parser now rejects previously valid parameterized activations in custom model YAMLs.

💬 Posted 2 inline comments

**Comment by glenn-jocher** [comment]
Trimmed the two compatibility risks we discussed:

- Removed the custom-dataset YAML location restriction, so dataset download scripts can still run from user YAMLs outside `ultralytics/cfg/datasets`.
- Replaced the AST activation parser with a compact constrained eval namespace that preserves existing activation strings (`torch.nn.*`, `nn.*`, and imported Ultralytics classes like `AGLU`) while removing builtins from the eval environment.
- Restored `check_is_path_safe()` behavior and kept future extraction-path validation local to archive extraction to avoid changing the public helper semantics.

Re-ran:
- strict Ruff on touched files
- `pytest tests/test_python.py -k "activation_literals or safe_download or grayscale"`
- `python -m compileall -q` on touched Python files

**Comment by glenn-jocher** [comment]
Trimmed this to a more practical 80/20 security PR:

- Reverted activation handling to existing behavior. We are not trying to solve the `eval()` finding in this PR because it is a public custom-model config escape hatch and prior fixes risked regressions.
- Kept custom dataset YAML `download:` scripts working outside `ultralytics/cfg/datasets`; no trusted-directory restriction.
- Restored `check_is_path_safe()` public semantics and kept non-existent extraction-path validation local to archive extraction.
- Removed the `AGLU` export/import/test churn.

Remaining focus is shared archive/path hardening plus small example/test scanner fixes. Re-ran strict Ruff, `pytest tests/test_python.py -k "safe_download"`, and compileall on touched files.

**Comment by UltralyticsAssistant** [comment]
Merged — fantastic work on this one, @glenn-jocher! 🚀

“An ounce of prevention is worth a pound of cure.” — Benjamin Franklin

This PR is a great example of that mindset in action: stronger archive and path safety, safer docs messaging, more reliable examples, and better test coverage all add up to a more secure and dependable experience for the Ultralytics community. Thoughtful improvements like these may look small in isolation, but they have a big impact across downloads, docs, examples, and developer trust.

Thank you for continuing to raise the bar for quality and safety in Ultralytics.
